src/front-end/typescript/lib/http/api/index.ts

```diff
@@ -1,3 +1,42 @@
+/**
+ * Exports all HTTP client functions to make requests to the
+ * Digital Marketplace back-end server. Each function is organized by
+ * resource and generally follows CRUD-naming conventions.
+ *
+ * @example
+ * Since this module re-exports all client functions, you should
+ * only import this module and not import the child modules directly.
+ * This module also re-exports some helper functions and types to deal
+ * with responses, like the {@link ResponseValidation} type.
+ * ```ts
+ * import * as api from "front-end/lib/http/api";
+ * import { Cmd } from "front-end/lib/framework";
+ * import { User } from "shared/lib/resources/user";
+ *
+ * type Msg = ADT<"gotUser", api.ResponseValidation<User, string[]>>;
+ *
+ * const getUserCmd: Cmd<Msg> =
+ *   api.users.readOne()("userId", response => adt("gotUser", response));
+ * ```
+ *
+ * @example
+ * Due to limitations with the TypeScript compiler, the client functions
+ * have been implemented inside "thunks" to enable better type inference.
+ * In practice, this means you may need to make an extra function call
+ * to use a client function, but this allows you to concretely set the type
+ * parameter for what the resulting {@link Cmd} returns.
+ * ```ts
+ * import * as api from "front-end/lib/http/api";
+ * import { User } from "shared/lib/resources/user";
+ *
+ * // Notice the <Msg> concrete type and the extra function call.
+ * const getUserCmd =
+ *   api.users.readOne<Msg>()("userId", response => adt("gotUser", response));
+ * ```
+ *
+ * @module lib/http/api
+ */
+
```

Comment:
This is more out of curiosity, but is there a reason we can't resolve the thunk in the call to create the client function?

For example, for reading a SWU proposal, could we do this? This would avoid having to make the extra function call before using it.

```ts
export function readOne<Msg>(
  opportunityId: Id,
  proposalId: Id,
  handler: HandleResponse<Resource.SWUProposal, string[], Msg> 
): component.cmd.Cmd<Msg> { 
  return crud.makeReadOneAction<RawSWUProposal, Resource.SWUProposal, string[], Msg>(
    NAMESPACE,
    rawSWUProposalToSWUProposal,
    `opportunity=${window.encodeURIComponent(opportunityId)}`
  )(proposalId, handler);
}
```

and then it could be used like this:

```ts
api.proposals.swu.readOne(opportunityId, proposalId, (response) =>
  api.isValid(response) ? response.value : null
),
```

Reply:
@sutherlanda Neat idea, and yes, that would work. I think the benefit of the thunk approach is that it avoids needing to enumerate the arguments and their type sigs for each client function (`proposalId` and `handler`). However, your suggested solution could result in less boilerplate overall. Do you think it's worthwhile to refactor the code and use this approach instead?

Reply:
Ah okay, yeah I'm not sure the benefit of less boilerplate is worth the tradeoff of having to enumerate all the arguments. I don't feel strongly either way, so I think it's good to keep as is if you're good with it.

Reply:
Same for me, I don't have a strong opinion on the matter either. Will leave it as-is. Thanks!